simulator/infra/cache/cache_tag_array.cpp

```diff
@@ -1,174 +1,94 @@
-    if ( !is_power_of_two( size_in_bytes / ( line_size * ways)))
-        serr << "ERROR: Wrong arguments! Number of sets should be a power"
-                  << " of 2." << critical;
-
-    if ( !is_power_of_two( line_size))
-        serr << "ERROR: Wrong arguments! Block size should be a power of "
-                  << "2." << critical;
+    return size_in_bytes / ( size_of_line * ways);
 }
 
-CacheTagArray::CacheTagArray( uint32 size_in_bytes,
-                              uint32 ways,
-                              uint32 line_size,
-                              uint32 addr_size_in_bits) :
-                              CacheTagArrayCheck( size_in_bytes, ways,
-                                                  line_size, addr_size_in_bits),
-                              num_sets( size_in_bytes / ( line_size * ways)),
-                              array( num_sets, std::vector<CacheTag>( ways)),
-                              lru( ways, num_sets)
-{
 
-}
 
 std::pair<bool, uint32> CacheTagArray::read( Addr addr)
 {
-    const auto lookup_result = read_no_touch( addr);
-    const auto&[ is_hit, way_num] = lookup_result;
+    uint32 num_set = set( addr);
+    uint32 num_tag = tag( addr);
 
-    if ( is_hit)
-    {
-        const auto set_num = set( addr);
-        lru.touch( set_num, way_num); // update LRU info
-    }
+    auto result = cache[ num_set].find( num_tag);
+    if ( result.first)
+        cache[ num_set].update( num_tag); // update LRU if it's a hit
 
-    return lookup_result;
+    return result;    
 }
 
+
+
 std::pair<bool, uint32> CacheTagArray::read_no_touch( Addr addr) const
 {
-    const auto set_num = set( addr);
-    const auto tag_num = tag( addr);
-
-    /* search into each way */
-    for ( uint32 i = 0; i < ways; ++i)
-    {
-        const auto& entry = array[ set_num][ i];
-
-        if ( entry.is_valid && entry.line == tag_num) // hit
-            return std::make_pair(true, i);
-    }
-    return std::make_pair(false, NO_VAL32); // miss (no data)
-}
+    uint32 num_set = set( addr);
+    uint32 num_tag = tag( addr);
 
-uint32 CacheTagArray::write( Addr addr)
-{
-    uint32 set_num = set( addr);
-    uint32 way_num = lru.update( set_num); // get l.r.u. way
+    return cache[ num_set].find( num_tag);
+}
 
-    array[ set_num][ way_num].line = tag( addr); // write it
-    array[ set_num][ way_num].is_valid = true; // this set is valid now
 
-    return way_num;
-}
 
-uint32 CacheTagArray::set( Addr addr) const
+uint32 CacheTagArray::write( Addr addr)
 {
-    /* Cut "logbin(line_size)" bits from the end. */
-    return (addr / line_size) & (num_sets - 1);
-}
+    uint32 num_set = set( addr);
+    uint32 num_tag = tag( addr);
 
-Addr CacheTagArray::tag( Addr addr) const
-{
-    /* Cut "logbin(line_size)" bits from the end. */
-    return addr / line_size;
-}
+    cache[ num_set].update( num_tag);
+    const auto&[ is_hit, value] = cache[ num_set].find( num_tag);
```

Comment:
Find is counter-intuition. Usually when we write data to the cache, we should find only a LRU way.